pulp_ansible/app/sigstore/tasks/sigstore_signature.py

```diff
@@ -0,0 +1,295 @@
+
+from pulp_ansible.app.models import (
+    CollectionVersion,
+    CollectionVersionSigstoreSignature,
+    SigstoreSigningService,
+)
+
+log = logging.getLogger(__name__)
+
+
+def _generate_checksum_manifest(collection_version):
+    """
+    Generate a file containing the sha256 hashes of all the collection elements.
+    """
+    cartifact = collection_version.contentartifact_set.select_related("artifact").first()
+    artifact_name = cartifact.artifact.file.name
+    artifact_file = storage.open(artifact_name)
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        differ = DistlibManifestChecksumFileExistenceDiffer
+        checksum = ChecksumFile(tempdir, differ=differ)
+        with tarfile.open(fileobj=artifact_file, mode="r") as tar:
+            tar.extractall(path=tempdir)
+
+            try:
+                manifest = checksum.generate_gnu_style()
+
+            except FileNotFoundError as e:
+                if os.path.islink(e.filename):
+                    log.error(
+                        f"Broken symlink found at {e.filename} -- this is not supported. Aborting."
+                    )
+                if e.filename.endswith("/MANIFEST.in"):
+                    log.error("Could not find a MANIFEST.in file in the specified project.")
+                    log.info("If you are attempting to sign a project, please create this file.")
+                    log.info("See the ansible-sign documentation for more information.")
+                raise e
+
+            except DistlibException as e:
+                log.error(f"An error was encountered while parsing MANIFEST.in: {e}")
+                raise e
+
+            for warning in checksum.warnings:
+                log.warn(warning)
+            log.debug(
+                "Full calculated checksum manifest (%s):\n%s",
+                manifest,
+            )
+            return manifest
+
+
+def verify_sigstore_signature_upload(data):
+    # Check that the provided signature materials correspond to the collection content on upload.
+    """The task code for verifying Sigstore signature upload."""
+    collection = data["signed_collection"]
+    repository = data.get("repository")
+    if repository:
+        sigstore_verifying_service = repository.sigstore_verifying_service
+    else:
+        raise ValueError("This content type must be associated with a repository.")
+    signature = data["data"]
+    certificate = data["sigstore_x509_certificate"]
+
+    def format(pubkey):
+        delimiter = "-----"
+        s = pubkey.split(delimiter)
+        res = delimiter + s[1] + delimiter + re.sub(" +", "\n", s[2]) + delimiter + s[3] + delimiter
+        return res
+
+    certificate = format(certificate)
+    sigstore_bundle = data.get("sigstore_bundle")
+    checksums = _generate_checksum_manifest(collection)
+
+    bundle = Bundle().from_json(sigstore_bundle) if sigstore_bundle != "null" else None
```

Comment:
Weird that a non-existant `sigstore_bundle` gets passed in as 'null' instead of `None` or simply not being present in `data`. Seems like a problem with the serializer field.

Reply:
Actually given the new standard adopted in sigstore-python (only the bundle is outputted by default as it already contains all the materials necessary for verifying the signature), I was thinking about removing the signature and certificate files from the `CollectionVersionSigstoreSignature` to simplify it. We can talk about it more in detail during our meeting if that's ok.

Reply:
What is the status of this now that V2 is out? Can this verification code be dramatically simplified? Can the verification service be removed entirely and be replaced with a text field like GPG?